### ISSUE TITLE
Property name (suffixes) differ after each generation of a proxy

### ISSUE THREAD

**Issue by holtkamp** [Ocramius/ProxyManager#367]
**TL;DR;** This is more of a question than an issue: is it really *required* to have a unique suffix of the property names each time they are generated?

Currently each time proxies are generated, the properties are slightly differently named, due to the use of the UniqueIdentifierGenerator (for example for the [$publicProperties](https://github.com/Ocramius/ProxyManager/blob/cce5477857504997baf3168974b8f1283516a686/src/ProxyManager/ProxyGenerator/PropertyGenerator/PublicPropertiesMap.php#L47)). Of course, functionally this is not a problem. But... is it required?

**Rationale**
I know proxies "should" not be committed to a Source Code Management system like GIT / SVN, but PaaS-providers like Heroku 'work' like this (with GIT). Generating the proxies is currently part of our build + deploy process, since a lot (around 900+, 40MB) are involved this is done before pushing the code.

Now on each deploy GIT detects changes in each of the proxies, while actually nothing has changed:

<img width="741" alt="screen shot 2017-03-14 at 10 56 31" src="https://cloud.githubusercontent.com/assets/776405/23896718/a86c449c-08aa-11e7-8e1f-1e63e773de50.png">

Possible options to circumvent this:
- have proxy generation take place after  it has been deployed, when the application starts
  - causes serious delays when starting up the application, significant spikes in response times
- have proxy generation take place as a post-install step in composer
  - at this point the autoload has already dumped
- have proxy generation take place as pre-autoload-dump step in composer
  - viable approach, could consider this
- have proxy generation take place when the code is pushed using [Heroku's 'compile' step](https://devcenter.heroku.com/articles/php-support#custom-compile-step)
  - too vendor specific?
- doubt why the property names change each time...
  - that is why I submitted this issue: why not make the suffix a hash of the classname + propertyName, why **should** it be unique? 
 
A subjective reason to prefer proxy generation before pushing, is the feeling to be 'in-control' about the content being pushed. When a proxy changes, this is clearly visible in the GIT history of Heroku application. Of course, having them change each time, kind of makes this useless.

**In summary**
So I guess my main question is: is it really required to have a unique suffix of the property names each time they are generated, would a hash suffice that only changes if really required?

**Comment by Ocramius** [comment]
> This is more of a question than an issue: is it really required to have a unique suffix of the property names each time they are generated?

Yes, because otherwise people will exploit those properties somewhere. What could be done is using a deterministic suffix, but it's simply not worth doing it, as the hashing parameters would need to cross many layers.

**Comment by holtkamp** [comment]
Aah, ok, yeah that is an argument 😄 thanks for the swift response.

**Comment by holtkamp** [comment]
>Yes, because otherwise people will exploit those properties somewhere.

Thinking about this again, but the visibility of these properties are all ```private```, so if someones tries to use them, of a proxy, with a crazy class name... well, that is really awkward 🤔 , right?

**Comment by Ocramius** [comment]
If someone is crazy enough to mess with this library's generated code, they
are crazy enough to do so via reflection API...

**Comment by holtkamp** [comment]
ok, but then 'supporting' this would not be this library's responsibility, right?

Feel free to close this issue, you got more experience in long-term maintenance of such project. I would say a "don't mess with or try to access the generated code" warning would suffice to cover any potential problems...

**Comment by Ocramius** [comment]
> ok, but then 'supporting' this would not be this library's responsibility, right?

I'm in a separate camp here: it should just not be possible at all.

**Comment by Ocramius** [comment]
@holtkamp I implemented this in https://github.com/Ocramius/ProxyManager/pull/385

Just note that there is a gotcha: the generated property name changes at every version change of your base repo.

This means that committing the proxies can effectively cause more proxy changes later.

Chicken-egg, but we are getting there, slowly. At least the name doesn't change each time you run the code generator :-)

**Comment by holtkamp** [comment]
Nice! One doubt:

>the generated property name changes at every version change of your base repo.

"my" base repo? [Seems to be the version of the ProxyManager](https://github.com/Ocramius/ProxyManager/pull/385/files#diff-51c12f174746e4e2d089de93e60e3a40R52), right?

**Comment by Ocramius** [comment]
@holtkamp `PackageVersions\Versions::VERSIONS` includes all versions of all dependencies, including your local project name.

**Comment by holtkamp** [comment]
Aah, ok, missed that, thanks!
